### ISSUE TITLE
Issue with AuthenticationAlgorithmHMACSHA256

### ISSUE THREAD

**Issue by lukeyeager** [gebn/bmc#49]
I am trying to adapt the example in `./cmd/describe/` to connect to a BMC using cipher suite 17. I had hoped that this diff would do the trick:
```diff
diff --git a/cmd/describe/main.go b/cmd/describe/main.go
index 4b02a03..af51e30 100644
--- a/cmd/describe/main.go
+++ b/cmd/describe/main.go
@@ -40,7 +40,7 @@ func main() {
 	ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
 	defer cancel()
 
-	machine, err := bmc.Dial(ctx, *argBMCAddr)
+	machine, err := bmc.DialV2(*argBMCAddr)
 	if err != nil {
 		log.Print(err)
 		return
@@ -72,10 +72,21 @@ func main() {
 		printSystemGUID(guid)
 	}
 
-	sess, err := machine.NewSession(ctx, &bmc.SessionOpts{
-		Username:          *flgUsername,
-		Password:          []byte(*flgPassword),
-		MaxPrivilegeLevel: ipmi.PrivilegeLevelUser,
+	sess, err := machine.NewV2Session(ctx, &bmc.V2SessionOpts{
+		SessionOpts: bmc.SessionOpts{
+			Username:          *flgUsername,
+			Password:          []byte(*flgPassword),
+			MaxPrivilegeLevel: ipmi.PrivilegeLevelUser,
+		},
+		AuthenticationAlgorithms: []ipmi.AuthenticationAlgorithm{
+			ipmi.AuthenticationAlgorithmHMACSHA256,
+		},
+		IntegrityAlgorithms: []ipmi.IntegrityAlgorithm{
+			ipmi.IntegrityAlgorithmHMACSHA256128,
+		},
+		ConfidentialityAlgorithms: []ipmi.ConfidentialityAlgorithm{
+			ipmi.ConfidentialityAlgorithmAESCBC128,
+		},
 	})
 	if err != nil {
 		log.Print(err)
```

However, when I try to use that, I get no response from the BMC for my first IPMI message:
```console
$ ./describe --username=X--password=X X
2022/04/27 13:46:45 connected to 10.X.X.X:623 over IPMI v2.0
ASF Presence Pong capabilities:
        IPMI:               true
        ASF v1.0:           true
        ASF security exts:  false
        DASH:               false
        DCMI:               true
Channel Authentication Capabilities:
        Channel:            0x1(Implementation-specific)
        Extended:           true
        SupportsV2:         true
        K_G configured:     false
        Per-message auth:   true
        User-level auth:    false
        Non-null usernames: true
        Null usernames:     false
        Anon login:         false
        OEM:                0(Unknown)
System:
        GUID:               X
2022/04/27 13:46:46 failed to get device id: read udp 10.X.X.X:35489->10.X.X.X:623: i/o timeout
```
I've tracked it down to an issue with the authentication algorithm. If I use `HMACSHA1` instead of `HMACSHA256` (while leaving the integrity and confidentiality algorithms unchanged), then I can connect just fine.

This is an issue because I am trying to use this library to connect to a new family of BMCs which requires the use of more modern ciphers. Currently I'm just falling back to my older, uglier solution of using `ipmitool` because when I use `ipmitool -I lanplus -C17` I can connect without issues.

**Comment by gebn** [comment]
Thanks for the report. The error suggests the session handshake succeeds, but the library attaches an invalid signature to the *Get Device ID* request, causing the BMC to drop the packet and not reply, resulting in the response timeout at our end.

Unfortunately I don't have access to a BMC supporting this authentication algorithm, so this could be awkward. As an initial sanity check, can you apply this patch, and check the output prints sane-looking 16-byte hashes:

```diff
diff --git a/pkg/ipmi/v2session.go b/pkg/ipmi/v2session.go
index 5638f8f..b4b29f8 100644
--- a/pkg/ipmi/v2session.go
+++ b/pkg/ipmi/v2session.go
@@ -3,8 +3,10 @@ package ipmi
 import (
        "crypto/hmac"
        "encoding/binary"
+       "encoding/hex"
        "fmt"
        "hash"
+       "log"
 
        "github.com/gebn/bmc/pkg/iana"
 
@@ -258,6 +260,7 @@ func (s *V2Session) SerializeTo(b gopacket.SerializeBuffer, opts gopacket.Serial
                }
 
                // append signature
+               log.Printf("appending %v-byte signature: %v", len(s.Signature), hex.EncodeToString(s.Signature))
                trailerSignature, err := b.AppendBytes(len(s.Signature))
                if err != nil {
                        return err
```

**Comment by lukeyeager** [comment]
Ah, so you want me to send you my improperly hashed password? I see through to your evil schemes! Just kidding, thanks for the pointer! Let me poke at it some more and let you know if I find anything.

**Comment by lukeyeager** [comment]
> Ah, so you want me to send you my improperly hashed password? I see through to your evil schemes!

As I'm sure you already knew, the signature you asked me to print doesn't seem to be related to the password - it comes after the session is already activated. So my joke probably fell flat.

> Unfortunately I don't have access to a BMC supporting this authentication algorithm

Ok. Presumably these commits from yesterday were related to you fixing error-checking when trying to use new ciphers on your old BMC, and aren't intended to help with me trying to connect to newer BMCs: https://github.com/gebn/bmc/compare/15ec63262b...7ed3d1504d

> As an initial sanity check, can you apply this patch, and check the output prints sane-looking 16-byte hashes:

The hashes look sane, I guess?
```
...
2022/04/28 08:47:23 appending 16-byte signature: a3250a8559f28b66f6ad8dd3a7067f4e
2022/04/28 08:47:24 failed to get device id: read udp 10.X.X.X:38757->10.52.2.21:623: i/o timeout
2022/04/28 08:47:24 appending 16-byte signature: 55a7d0344f2fa42d058e19c7307902c5
2022/04/28 08:47:25 failed to get chassis status: read udp 10.X.X.X:38757->10.52.2.21:623: i/o timeout
...
```
I confess I don't understand why we're looking at a signature which is generated by the IntegrityAlgorithm, since it seems to me that the problem is with the AuthenticationAlgorithm?

From this section of code (https://github.com/gebn/bmc/blob/7ed3d1504d/v2session_new.go#L170-L234):
* authAlgo is used to create the SIK
* the SIK is used when creating `keyMaterialGen` which is an input to `hasher`
* `hasher` is used as the `integrityAlgorithm` for the V2Session object

So, I do see a trail from the authAlgo to the integrityAlgo, but I don't understand why the rakpMessage3 and 4 handshake seems to work fine (which uses authAlgo to generate the authCode), but then the first message fails later which also tries to use this authAlgo. I would suspect that you accidentally used used authAlgo instead of integrityAlgo somewhere (or vice versa), except that the library works when using other cipher suites.

I know you don't have access to hardware for debugging. I'm game to try and create a patch, but I'm struggling to line up the IPMI v2 spec with your code to try to spot an issue.

**Comment by lukeyeager** [comment]
Any tips for debugging? I have .pcap files from `ipmtool -C17` vs this library, but I don't immediately see anything wrong. If I had something like https://github.com/gebn/bmc/issues/5, then maybe I could spot the difference more easily.

**Comment by gebn** [comment]
Hashes look good, however it's strange how there is no evidence of retries. What's the latency between client and BMC? It could be worth upping the response timeout with something like `machine.SetTimeout(3 * time.Second)`. The default is 1 second.

`User-level auth` being false is the BMC saying it will allow unsigned "read-only" commands. This library signs everything, so I wondered if the BMC could be rejecting packets with a redundant signature, even if the spec says this is allowed. IPMItool does the same (albeit when I was using a different cipher suite), so I think that's a dead end.

Ultimately I think this comes down to a comparison of input and output with IPMItool. Can you run:

```
ipmitool -vvv -I lanplus -H <ip> -U <username> -P <password> raw 0x06 0x01
```

It'll send a whole bunch of commands. The bits I need are the `Generated K1 (20 bytes)`, then under the penultimately built command (netfn 0x06 and command 0x01), the `authcode input (48 bytes)` and `authcode output (16 bytes)`.

Please then apply the following patch, which will give you the equivalent 3 fields from this library. They will be different due to various random numbers, but I'm interested in where they differ.

```diff
diff --git a/pkg/ipmi/v2session.go b/pkg/ipmi/v2session.go
index 5638f8f..ec5566b 100644
--- a/pkg/ipmi/v2session.go
+++ b/pkg/ipmi/v2session.go
@@ -3,8 +3,10 @@ package ipmi
 import (
        "crypto/hmac"
        "encoding/binary"
+       "encoding/hex"
        "fmt"
        "hash"
+       "log"
 
        "github.com/gebn/bmc/pkg/iana"
 
@@ -255,6 +257,8 @@ func (s *V2Session) SerializeTo(b gopacket.SerializeBuffer, opts gopacket.Serial
                // if requested, calculate signature
                if opts.ComputeChecksums {
                        s.Signature = executeHash(s.IntegrityAlgorithm, b.Bytes())
+                       log.Printf("auth code input (%v): %v", len(b.Bytes()), hex.EncodeToString(b.Bytes()))
+                       log.Printf("auth code output (%v): %v", len(s.Signature), hex.EncodeToString(s.Signature))
                }
 
                // append signature
diff --git a/v2session_new.go b/v2session_new.go
index 140116d..507e6e9 100644
--- a/v2session_new.go
+++ b/v2session_new.go
@@ -7,6 +7,7 @@ import (
        "encoding/hex"
        "errors"
        "fmt"
+       "log"
 
        "github.com/gebn/bmc/pkg/ipmi"
 
@@ -196,6 +197,7 @@ func (s *V2SessionlessTransport) newV2Session(ctx context.Context, opts *V2Sessi
        }
        sikHash := hashGenerator.SIK(effectiveBMCKey)
        sik := calculateSIK(sikHash, rakpMessage1, rakpMessage2)
+       log.Printf("SIK: %v", hex.EncodeToString(sik))
        icvHash := hashGenerator.ICV(sik)
        rakpMessage4ICV := calculateRAKPMessage4ICV(icvHash, rakpMessage1,
                rakpMessage2)
```

**Comment by lukeyeager** [comment]
I'm seeing `Generated K1 (32 bytes)` - not 20.
```console
$ ipmitool ...
...
Generated K1 (32 bytes)
 01 7c c3 be 77 5c a7 1e 3b 6d 56 0c 98 6a a6 50
 40 2e d3 b1 c7 f5 af 0b 6b 26 a8 3c f2 6c 84 57
...
authcode input (48 bytes)
 06 c0 ec c1 e8 db 06 00 00 00 20 00 87 74 b2 f3
 d1 e3 f0 79 70 7c 85 5a d0 17 c9 fb e5 83 3b fe
 af d4 9d 3c 2b b2 59 87 bb 55 81 aa ff ff 02 07
authcode output (16 bytes)
 da c3 9d aa 45 07 9c 9f c1 28 4b 50 6f 8a 1e 88
...
>> Sending IPMI command payload
>>    netfn   : 0x06
>>    command : 0x01
>>    data    :

$ ./describe ...
2022/05/02 14:33:15 SIK: 073008d9a3002d468068603f30aeb2d7de188b25bd4ef30ebb4b7349910801a7
2022/05/02 14:33:15 auth code input (48): 06c0fcbb0be90100000020001c60f0444fa0dc576f6037c48b99ecc747457df963dd2897c288a6bfc8a8fe1cffff0207
2022/05/02 14:33:15 auth code output (16): 9c7bd8c3b0a5fa76246ae112708dd063
```
I've sent you the full logs via email just so I don't accidentally fail to strip something confidential. Feel free to respond here

**Comment by gebn** [comment]
Ok. Can you check out a clean copy of `master`, apply this patch, then try your `./describe` again:

```diff
diff --git a/authenticator.go b/authenticator.go
index 3316faf..7cc824e 100644
--- a/authenticator.go
+++ b/authenticator.go
@@ -144,8 +144,8 @@ func (g additionalKeyMaterialGenerator) K(n int) []byte {
        // when the spec says HMAC block size, it means the size of the output tag,
        // not the block size of the underlying algorithm (e.g. 20 rather than 64
        // for SHA-1).
-       constant := make([]byte, g.hash.Size())
-       for i := 0; i < g.hash.Size(); i++ {
+       constant := make([]byte, 20)
+       for i := 0; i < 20; i++ {
                constant[i] = uint8(n)
        }
        return executeHash(g.hash, constant)
```

**Comment by lukeyeager** [comment]
Wow, that did the trick, thanks! If you're inclined to explain what you've learned, I'd love to know! Otherwise, thanks for being a top-notch open-source contributor.

Should I just plan to run a version patched with `s/g.Hash.Size()/20/`, or can you find a way to integrate it into the code?

**Comment by gebn** [comment]
Glad it worked! Thanks for the extra output in your email - that was necessary to hunt it down. The smoking gun was a different K1 to IPMItool given the same SIK, which should never happen. Initially I tried debugging with IPMItool's K1, which of course made everything work, so the lesson there is to not assume the place the bug manifests itself is where the problem is. That, and don't trust the spec.

Patch is already integrated. `RAKP-HMAC-SHA1` worked because `g.hash.Size()` happened to be 20 with that algorithm, so it's backwards-compatible. I've documented in more detail [in the code](https://github.com/gebn/bmc/blob/b6efe2f31ce0381fd9b4fd3516c45405019b4578/authenticator.go#L144-L148).
